### PR TITLE
XOR Reorganized for Error Reporting (Work in Progress)

### PR DESCRIPTION
**Pull request by AE1020**
This is a prototype that picks one operator to work with...to see about bubbling up the necessary information to report errors ( as discussed in #174 ) and tries to make some other improvements in the process.

Truly getting to a point where out of memory can be handled gracefully will require work and stress testing--but this tries to put things on that path.

Part of the goal is to streamline the dispatching to be more regular--so that the subdispatch matches the interface of the higher-level dispatch.  This reduces the amount of transformation that has to be done.

For instance: there were boolean return codes which had to be interpreted by callers as a container type.  This gives the subdispatch the place to write the container type if it changes--and it can leave it as-is if not.  <strike>Ultimately this might be imagined as updating the container and the type in the container array directly...without needing separate code to update it after the dispatch returns.</strike> **(done)**

**Stylistic Notes**

Writing out the full word "container" is less necessary when it's in the type (`container_t *c` is better than `void *container`), and most of the functions redundantly have container in their name anyway.

Out of the abbreviations for container classes (`array/run`, `arr/run`, `ac/rc`) this favors `c` for generic containers and `ac`/`rc`/`bc` for specific ones.  This communicates while being brief, and `ac->array` looks better than `array->array` or `arr->array`, with the A.C. reinforcing the concept that it is an "Array *Container*".

<strike>The bitset container's inner words array would be better called `words`, as having it be called "array" is a bit of a speedbump when working with containers where one is an "array container" and one is a "bitset container".</strike> [merged in a separate PR](https://github.com/RoaringBitmap/CRoaring/pull/269)

When there are two inputs they are numbered and labeled by their type, since **ac1** and **rc2** offers strictly more information than **src1** and **src2**.

### CODE REVIEW

**Comment by AE1020** [comment]
So this is a direction that I think is good.  Awaiting review.

It flattens the protocols in a way that makes it easier to understand, and easier to add systemic improvements.

One such improvement is having a unified way to bubble up results.

It also enabled adding an optimization to take advantage of already being in the "mixing" routine and knowing what kind of container is being returned.  There's no need to re-switch() on a container type to call the right test for if it's empty, or to free it.  This could also mean avoiding making any container at all.

File: src/roaring.c [hunk 1]
```diff
@@ -2379,51 +2307,50 @@ roaring_bitmap_t *roaring_bitmap_lazy_xor(const roaring_bitmap_t *x1,
...
             container_t *c2 = ra_get_container_at_index(
                                     &x2->high_low_container, pos2, &type2);
  
-            // We do the computation "in place" only when c1 is not a shared container.
-            // Rationale: using a shared container safely with in place computation would 
-            // require making a copy and then doing the computation in place which is likely 
-            // less efficient than avoiding in place entirely and always generating a new 
-            // container.
+            if (!(*dispatcher)(c1, type1, c2, type2))
+                goto alloc_failure;  // c1 should still be freeable
```
Comment (lemire) [inline src/roaring.c:2348]: I recommend using curly braces after an if clause *unless* the if clause fits on a single line.

File: src/roaring.c [hunk 2]
```diff
@@ -2447,8 +2375,12 @@ void roaring_bitmap_lazy_xor_inplace(roaring_bitmap_t *x1,
                 ra_set_container_at_index(&x2->high_low_container, pos2, c2,
                                           type2);
             }
-            ra_insert_new_key_value_at(&x1->high_low_container, pos1, s2, c2,
-                                       type2);
+
+            if (!ra_insert_new_key_value_at(&x1->high_low_container,
+                                            pos1, s2, c2, type2))
+            {
```
Comment (lemire) [inline src/roaring.c:2381]: The current style used throughout is, I believe to have the curly brace on the same line as the "if".

File: include/roaring/roaring_array.h [hunk 3]
```diff
@@ -72,6 +72,19 @@ void ra_clear_without_containers(roaring_array_t *r);
...
+ * Make it so that a crash is likely if a container is used after an operation
+ * partially ran, generating undefined results.
+ *
+ * (API users should only encounter such states when using core versions
+ * that return error conditions vs. panic.  So it would be redundant to add
+ * checks for a "corrupt flag" on each API call that would panic if set.)
+ */
+inline static void ra_mark_corrupt(roaring_array_t *ra) {
+    assert(ra->containers == NULL);  // all containers must be free
+    ra->containers = (container_t**)((intptr_t)0xDECAFBAD);
```
Comment (lemire) [inline include/roaring/roaring_array.h:85]: This seems odd. Is this something we find in other major important code bases?

File: src/roaring.c [hunk 4]
```diff
@@ -2447,8 +2375,12 @@ void roaring_bitmap_lazy_xor_inplace(roaring_bitmap_t *x1,
...
+                goto alloc_failure;
```
Comment (lemire) [inline src/roaring.c:2382]: Is the goto genuinely useful here?

File: src/roaring.c (hunk 1 above)
Comment (lemire) [inline src/roaring.c:2348]: It seems that the goto is not genuinely needed.

**Comment by lemire** [comment]
I like it a lot. It might be useful to formally document the "safe usage".

File: src/roaring.c (hunk 1 above)
Comment (AE1020) [reply src/roaring.c:2348]: I never put a branch on the same line--because it can be extremely annoying when you want a breakpoint on the branch.  (While you can in some IDEs find a way to do it in the assembly view, the space savings is never worth having to do that.)  I also find it easier to overlook a branch on the same line.  The indentation calls out "hey, branch here!"

My opinion on braces for single-line conditions is that if modern compilers didn't have good warnings for catching unintended consequences, I'd certainly be for using them.  But todays warnings are good enough, so I find them to be visual noise.  This is not a huge deal to me.

*The main thing I think RE: formatting* is that it's a hassle whenever you require adjustments and alignments each time a name length changes.  Maybe it's appropriate for public-facing API headers that are going for a certain layout and look to be taken in more easily.  But for day-to-day code, I prefer:

     void some_function_name(
         int some_argument_name,
         int some_other_argument,
         int yet_another_argument
     ){
         /* function content here */
     }
     
...over more acrobatic forms which push things way off to the right, and require more maintenance and cognitive tax each time you want to make an adjustment.

Comment (lemire) [reply src/roaring.c:2348]: I am not very interested in style issues, but I am sure we agree that one should be reasonably consistent so that the whole code base has a somewhat uniform style. That's all I am saying.

I do not advocate anything more.

File: include/roaring/roaring_array.h (hunk 3 above)
Comment (AE1020) [reply include/roaring/roaring_array.h:85]: Depending on what you want to call precedent... there's the "blue screen of death".  Anything generally in the notion that it's better to trigger things to stop or crash than to continue on with invalid data.

I mentioned [the three possibilities here](https://github.com/RoaringBitmap/CRoaring/issues/174#issuecomment-721523782).

As it happens, I've had to add a flag for this state so the C++ destructor can run the clear() function.  But the flag isn't paid for on checking at every entry point.  One could do so and then error, but that seems an unnecessary expense for this edge case.

Comment (lemire) [reply include/roaring/roaring_array.h:85]: I am not arguing the purpose, I am asking about how you reassign it from NULL to 0xDECAFBAD. I have never seen this strategy before and I am not sure I understand the reasoning behind it.

Comment (AE1020) [reply include/roaring/roaring_array.h:86]: If it were left as NULL it would be a valid empty container, and this is to make it invalid.

If you are not familiar with 0xDECAFBAD, it's just a traditionally used "easy to see in hex dumps" garbage value.

https://nedbatchelder.com/text/hexwords.html

Comment (lemire) [reply include/roaring/roaring_array.h:86]: @AE1020  I know what hex words are... but I am still puzzled. 

Can you give a user-facing example where this would protect the user? Presumably, you have in mind a case where if we did not do this assignment, then something bad would happen.

I am not seeing the scenario in question.

In effect, I would think that we don't want to do something in the hope that it might be protective. We want to do something that we can test out and see that it does protect us.

Comment (AE1020) [reply include/roaring/roaring_array.h:86]: The bad thing is believing that your operation (in this case, an XOR) produced an empty result, when there was a memory allocation failure.

    if (!roaring_bitmap_xor_inplace_completed(x1,x2)) {
        // memory must be low, ask the GC to reclaim space
        garbage_collect();
        // that may take a while, run interrupts to check for cancellation
        run_interrupts();
        ....
    }

The question is if there is some amount of forgetfulness in the fact that x1 doesn't represent anything meaningful.  If you forget, I say it's better to raise the odds of crashing than continuing, if there are cheap ways of doing so.

Now, what is actually checked for meaningfulness is the ROARING_FLAG_INDETERMINATE:

https://github.com/RoaringBitmap/CRoaring/pull/267/commits/28ec27cde685ffff35bebfb6890b278b42705e29#diff-0e67f5a62cae4679dbfed795bee9a7f80a7195f23cf7c1b7b4f340064e08e581R424

If this edge case were addressed "rigorously" then you would at each API entry point make sure you weren't dealing with ROARING_FLAG_INDETERMINATE (which was added later).  I deemed that overkill for this--but I don't think it's a good idea to treat it as an empty set like nothing has happened.

You either treat it as an empty set, pay for flag checks at a lot of entry points, or use the inexpensive but effective option of throwing it a curveball.  Out of the three, that would be my choice.  You might be able to do it other ways that force null dereference (e.g. giving it a large size, leaving the pointer NULL).  The "idiom" of a trash hex pointer is just simple and familiar.

File: src/roaring.c (hunk 1 above)
Comment (AE1020) [reply src/roaring.c:2388]: These are works in progress of a line of thinking, better seen in a more current example:

https://github.com/RoaringBitmap/CRoaring/blob/d90e7678fdc3e7ca672bea0fffcf62b03b09ee6b/src/roaring.c#L962

I've intentionally made the branches there use different styles for purposes of contrast.  The `s1 > s2` branch is more typically in line of my preferences of how I like C code to look...to me that is easier to scan.

I find this to be an effective use of gotos.

File: include/roaring/roaring_array.h (hunk 3 above)
Comment (lemire) [reply include/roaring/roaring_array.h:86]: So in this example...

```C++
if (!roaring_bitmap_xor_inplace_completed(x1,x2)) {
    // memory must be low, ask the GC to reclaim space
    garbage_collect();
    // that may take a while, run interrupts to check for cancellation
    run_interrupts();
    ....
}
```

The function has made x1 invalid. You want that, with some probability, access to `x1` will crash. I don't think that assigning a crazy address ensures that this address plus some offset will crash. You are just pretty sure that it will crash because the memory space is vast, and you are unlikely to be unlucky enough that your fixed address plus some offset will end up overlapping a valid memory address.

Why not just take a pointer to x1 and reassign x1 to NULL. Then the reference is gone for good and you are done with it...

```C++
if (!roaring_bitmap_xor_inplace_completed(&x1,x2)) {
  // here x1 == NULL
    ....
}
```

Comment (AE1020) [reply include/roaring/roaring_array.h:86]: > Why not just take a pointer to x1 and reassign x1 to NULL.

That might be sensible if you were guaranteed that the pointer you're being passed was one you could free and clobber the client's copy.

...but...the model you're working with is that you have a non-opaque struct that can be resident in client-owned memory, that may or may not be dynamically allocated (e.g. the C++ object doesn't do an allocation)

And if someone uses their own allocator...and you nulled out their pointer in this process...they'd have no way to free it.

> You want that, with some probability, access to x1 will crash. I don't think that assigning a crazy address ensures that this address plus some offset will crash

While it doesn't ensure it...it's a defensive move that covers most all cases.  Any formal solution for how to handle an invalid block of client-owned memory will have an associated cost; which (again) I do not think this case warrants.  I just consider it better than doing nothing.

Comment (AE1020) [reply include/roaring/roaring_array.h]: I realized that the name is somewhat triggering, as its from a historical path where the need to be able to tell whether the structure was in this state had not yet been absorbed.

[This renames it from ra_mark_corrupt() to ra_mark_indeterminate()](https://github.com/RoaringBitmap/CRoaring/pull/267/commits/6b0a370ca5af2ba90dc7df618880b8686939e0f9), and changes some of the points of emphasis.

Important to remember in the malloc()'-ing and free()'-ing world of C, the aren't guarantees when you break contracts.  Dereferencing NULL also doesn't *guarantee* a crash:

[https://stackoverflow.com/questions/6793262/why-dereferencing-a-null-pointer-is-undefined-behaviour](https://stackoverflow.com/questions/6793262/why-dereferencing-a-null-pointer-is-undefined-behaviour)

So even if you were to start out the process by allocating a global "bigalloc" for a single container, and then set the container inside that to NULL, and use that for every indeterminate struct... it's just additional work that obscures what's going on.

**Comment by AE1020** [comment]
I've managed to run a burn-in test overnight with sporadic random allocations failing that runs "correctly"--e.g. doesn't leak, doesn't crash, lets you know your XOR ran out of memory and you can continue.  (I've now got some checkpointing in the random test to make it easier to get back to a point to debug if it only occurs at a very late iteration.)  So it seems good.

This took a while...and propagating it to *all* the operations would be a *significant* amount of work.  It's also not as easy to homogenize the approach for operations that have more custom in-place operations; I have to look at that more closely.  Plus since the tests are tied closely to the internals, changing parameterization of basically any internal has to pay the tax of the tests too.
 
Since it's mapped out and you've had a chance to review it, I'm going to close the PR for now, rather than leave it hanging.  I'll see about picking the branch up again when I have time to devote to it.  I may have to make more modest tactical changes to meet my immediate needs...

**Comment by lemire** [comment]
@AE1020 Thank you. I think that this is great work. Looking forward to reading you again.

File: include/roaring/roaring_array.h (hunk 3 above)
Comment (lemire) [reply include/roaring/roaring_array.h]: Seems sensible.
